Add CDC partitioner (version 3.10.2)

avelanarius opened the pull request:
Add new class `CDCToken` which implements a Scylla CDC partitioner, as defined here:

https://github.com/scylladb/scylla/blob/master/cdc/cdc_partitioner.cc

Modify `Statement` classes to provide table name alongside already implemented keyspace name, in order to detect if CDC partitioner should be used for a given `Statement`.

Modify `Metadata` to use CDC partitioner instead of cluster-wide partitioner when calculating token for CDC log table. As Scylla token range information is shared between Murmur3 partitioner and CDC partitioner only a single `TokenMap` is used. 

(In case of Scylla) this TokenMap stores `M3PToken` tokens, but queries for `CDCToken` are allowed - `M3PToken` and `CDCToken` are modified to allow comparison between those types (as they are both represented by 64-bit integers).

### Manual testing
I have setup a table with CDC enabled on a 2-node Scylla cluster. Exectued a few prepared statements using Java client, reading from CDC log table with `WHERE` clause on partition key (`cdc$stream_id`) and enabled tracing.

The tracing shows the change works properly (click to enlarge):

![tracings](https://user-images.githubusercontent.com/10003183/93330169-aa69f700-f81e-11ea-9b5a-cd7aca167150.png)

haaawk commented on driver-core/src/main/java/com/datastax/driver/core/BatchStatement.java:
```diff
@@ -284,6 +284,15 @@ public String getKeyspace() {
 
+  @Override
+  public String getTable() {
+    for (Statement statement : statements) {
```
This is controversial. We should return the table name only if every statement in a batch targets the same table. Otherwise, let's return null.

haaawk commented on driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java:
```diff
@@ -388,7 +388,12 @@ private Connection findLeastBusyForShard(int shardId) {
 
   ListenableFuture<Connection> borrowConnection(
-      long timeout, TimeUnit unit, int maxQueueSize, ByteBuffer routingKey) {
+      long timeout,
```
This is the wrong style. The project does not break lines like this.

avelanarius replied:
The Maven project has `fmt-maven-plugin` installed, which checks the style and allows you to automatically format the code (this is how this tool formatted it). I will try to reformat the code, so that both automatic checker and project style are ok.

haaawk replied:
If the automatic checker works then it's ok and don't mind my comment.

haaawk commented on driver-core/src/main/java/com/datastax/driver/core/RequestHandler.java:
```diff
@@ -98,13 +98,14 @@
     if (keyspace == null) {
       keyspace = loggedKeyspace;
     }
+    String table = statement.getTable();
```
better to define the variable just before it's used.

haaawk commented on driver-core/src/main/java/com/datastax/driver/core/SessionManager.java:
```diff
@@ -733,7 +733,13 @@ public void run() {
                 .getValue()
-                .borrowConnection(0, TimeUnit.MILLISECONDS, 0, statement.getRoutingKey());
+                .borrowConnection(
+                    0,
```
again, I don't think this is the style of the project. Correct me if I'm wrong.

haaawk commented on driver-core/src/test/java/com/datastax/driver/core/QueryLoggerTest.java:
```diff
@@ -320,6 +320,11 @@ public String getKeyspace() {
           }
 
+          @Override
+          public String getTable() {
```
This is not needed, is it?

haaawk commented on driver-core/src/main/java/com/datastax/driver/core/querybuilder/BuiltStatement.java:
```diff
@@ -103,8 +104,12 @@
 
   BuiltStatement(
-      String keyspace, List<ColumnMetadata> partitionKey, List<Object> routingKeyValues) {
+      String keyspace,
```
Wrong style again.

haaawk commented:
A general comment:
I think it would be better for a statement to have isCDC() instead of getTable(). This way the check we have to do on every execution of the statement would be cheaper.

haaawk commented:
> A general comment:
> I think it would be better for a statement to have isCDC() instead of getTable(). This way the check we have to do on every execution of the statement would be cheaper.

The most general way of implementing this would be to give each statement a getPartitioner() method which would return the token factory which should be used for this statement.

haaawk commented:
Also, we should try to reduce the number of files/classes affected by this patch. This will make it easier for us to rebase it on top of updated upstream branches.

I would suggest:
1. Addition of getTokenFactory()/getPartitioner() to Statement with default implementation that returns null
2. Implement the new method only for BoundStatement. Every other Statement implementation can use the default version of the method which returns null.

haaawk commented on driver-core/src/main/java/com/datastax/driver/core/TableOptionsMetadata.java:
```diff
@@ -183,8 +184,10 @@
-    if (is380OrHigher) this.cdc = isNullOrAbsent(row, CDC) ? DEFAULT_CDC : row.getBool(CDC);
-    else this.cdc = DEFAULT_CDC;
+    boolean hasCDCExtension = this.extensions.containsKey(CDC_EXTENSION);
+    if (is380OrHigher)
```
We need separate field. this.cdc is a Cassandra CDC. The driver has to work with both Cassandra and Scylla so we need something like this.scyllaCDC as Cassandra CDC and Scylla CDC are separate concepts.

avelanarius replied:
Fixed in https://github.com/scylladb/java-driver/pull/35/commits/16bfd04695707cf42ef3b9be347d1a5010aa23cf

haaawk commented on driver-core/src/main/java/com/datastax/driver/core/Token.java:
```diff
@@ -671,4 +679,117 @@ public String toString() {
+              "CDC partition key has invalid length: expected {} bytes, but got {} bytes",
+              CDC_PARTITION_KEY_LENGTH,
+              length);
+          return MIN_TOKEN;
```
This may cause a disaster because all keys will end up with the same token. It would be better to take the first 8 bytes of the key if they are present. If length < 8 then add 0 to the front. It's important that key -> token mapping was stable otherwise the user won't be able to find the data so random token won't work.

avelanarius replied:
It is implemented the same way as in Scylla:

https://github.com/scylladb/scylla/blob/a81b731a1a2f5a07b0b5b8fde7939be005be0a97/cdc/cdc_partitioner.cc#L42-L48

However I see the issue and I will fix it.

haaawk replied:
Fair point. It might be worth fixing there as well.

avelanarius replied:
Fixed in https://github.com/scylladb/java-driver/pull/35/commits/68408434963e1ba34183c397e5d198558ed489bc

avelanarius committed:
Add CDC partitioner

Add new class CDCToken which implements a Scylla
CDC partitioner, as defined here:

https://github.com/scylladb/scylla/blob/master/cdc/cdc_partitioner.cc

Add new getPartitioner() method in Statement
in order to be able to use custom one on CDC log
tables (and potentially other in the future).

Modify DefaultPreparedStatement to set
CDCPartitioner for CDC log prepared statements.

Change Metadata to be able to use custom
partitioner (or cluster-wide if null is
provided).

As Scylla token range information is shared
between Murmur3 partitioner and CDC partitioner
only a single TokenMap is used.

(In case of Scylla) this TokenMap stores
M3PToken tokens, but queries for CDCToken
are allowed - M3PToken and CDCToken are modified
to allow comparison between those types (as they
are both represented by 64-bit integers).

avelanarius commented:
@haaawk In the amended commit:

1. Removed `getTable()` and added `getPartitioner()` (as you described).
2. Introduced Scylla-specific `isScyllaCDC()` (`isCDC()` is kept the same as before this PR).
3. Did NOT change CDC hash behaviour when partition key has invalid length - still returning `MIN_TOKEN`.

I experimented with the automatic code formatter, but it seems you can't avoid long method arguments split in many lines instead of one (checker doesn't allow other formattings).

haaawk commented:
The change looks good know. Please change the CDC hash behavior and we will be ready to merge. The behavior should be something like this:
```
if (length < 8) pick MIN_TOKEN
if (length != 16) use the first 8 bytes as token and don't check the version
otherwise do the normal logic
```

avelanarius commented:
@haaawk Implemented your suggested CDC hash behaviour. The code should be ready for merge.

haaawk reviewed (approved):
LGTM Thanks
